extensions/replication/tasks/MultipleBackendTask.js

```diff
@@ -343,26 +381,21 @@ class MultipleBackendTask extends ReplicateObject {
      * @return {Array} The array of byte ranges.
      */
     _getRanges(contentLen, isGCP) {
-        // 5MB min part size rounded up to the nearest power of two is 8388608.
-        let partCount = Math.pow(2,
-            Math.ceil(Math.log(contentLen / 8388608) / Math.log(2)));
-        // The GCP storage type does not accept an MPU that is > 1024 parts, so
-        // we perform an MPU of <= 1024 parts if replicating to a GCP location.
-        // Otherwise, we use a max part count of 8192 which is the AWS max part
-        // count of 10000 rounded down to the nearest power of two. The AWS max
-        // part size is 5GB, so an MPU of 8192 parts allows for reaching the AWS
-        // max object size of 5T.
-        const maxPartCount = isGCP ? MPU_GCP_MAX_PARTS : 8192;
-        if (partCount > maxPartCount) {
-            partCount = maxPartCount;
-        }
-        const pow = Math.pow(2, Math.ceil(Math.log(contentLen) / Math.log(2)));
-        const range = Math.ceil(pow / partCount);
+        let size = isGCP ?
+            this._getGCPRangeSize(contentLen) : this._getRangeSize(contentLen);
         const ranges = [];
         let start = 0;
         let end = 0;
         while (end < contentLen - 1) {
-            end = start + range - 1;
+            // If the object is large enough to reach 9760 parts of 512MB, then
+            // at this point we need to increase the part size to 1GB, such that
+            // the largest object size of 5TB can be accounted for by performing
+            // a 10,000 part multipart upload. That is, 9760 parts of size 512MB
+            // and 240 parts of size 1GB is 5TB.
+            if (ranges.length === 9760 && size === 512 * (1024 * 1024)) {
```

Comment:
I think this logic is brittle and not really needed, I would rather have the same part size for all to avoid surprises. Hence if you need 1GB parts because of the object being close to 5TB, you could add this in `_getRangeSize` before `return rangeSize`:

```js
// make sure we do not need more than 10K parts for the biggest allowable objects
while (contentLen / rangeSize > 10000) {
    rangeSize *= 2;
}
```

Reply:
Great, I like the simplification! The goal here was to allow us the potential to use up to 10K parts, but I'm not wedded to that. Potentially using up to 9760 parts of 512MB each seems good enough to me.